pkg/kubelet/cm/cgroup_manager_linux.go

```diff
@@ -53,74 +53,71 @@ const (
+		// names to systemd cgroup driver, we want to remap "-" => "_",
+		// so we forbid "_" so that we can always revert the mapping.
+		if strings.Contains(component, "/") || strings.Contains(component, "_") {
+			panic(fmt.Errorf("Invalid character in component [%s] of CgroupName", component))
 		}
-		// each part is appended with systemd style -
-		result = strings.Join(results, "-")
-	} else {
-		// root converts to -
-		result = "-"
 	}
-	// always have a .slice suffix
-	if !IsSystemdStyleName(result) {
-		result = result + systemdSuffix
+	return CgroupName(append(base, components...))
+}
+
+// cgroupName.ToSystemd converts the internal cgroup name to a systemd name.
+// For example, the name {"kubepods", "burstable", "pod1234-abcd-5678-efgh"} becomes
+// "/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod1234_abcd_5678_efgh.slice"
+// This function always expands the systemd name into the cgroupfs form. If only
+// the last part is needed, use path.Base(...) on it to discard the rest.
+func (cgroupName CgroupName) ToSystemd() string {
+	if len(cgroupName) == 0 || (len(cgroupName) == 1 && cgroupName[0] == "") {
+		return "/"
+	}
+	newparts := []string{}
+	for _, part := range cgroupName {
+		part = strings.Replace(part, "-", "_", -1)
+		newparts = append(newparts, part)
 	}
 
-	// if the caller desired the result in cgroupfs format...
-	if outputToCgroupFs {
-		var err error
-		result, err = cgroupsystemd.ExpandSlice(result)
-		if err != nil {
-			panic(fmt.Errorf("error adapting cgroup name, input: %v, err: %v", name, err))
-		}
+	result, err := cgroupsystemd.ExpandSlice(strings.Join(newparts, "-") + systemdSuffix)
+	if err != nil {
+		// Should never happen...
+		panic(fmt.Errorf("Error converting cgroup name [%v] to systemd format: %v", cgroupName, err))
 	}
 	return result
 }
 
-// ConvertCgroupFsNameToSystemd converts an expanded cgroupfs name to its systemd name.
-// For example, it will convert test.slice/test-a.slice/test-a-b.slice to become test-a-b.slice
-// NOTE: this is public right now to allow its usage in dockermanager and dockershim, ideally both those
-// code areas could use something from libcontainer if we get this style function upstream.
-func ConvertCgroupFsNameToSystemd(cgroupfsName string) (string, error) {
-	// TODO: see if libcontainer systemd implementation could use something similar, and if so, move
-	// this function up to that library.  At that time, it would most likely do validation specific to systemd
-	// above and beyond the simple assumption here that the base of the path encodes the hierarchy
-	// per systemd convention.
-	return path.Base(cgroupfsName), nil
+func ParseSystemdToCgroupName(name string) CgroupName {
+	driverName := path.Base(name)
+	driverName = strings.TrimSuffix(driverName, systemdSuffix)
+	parts := strings.Split(driverName, "-")
+	result := []string{}
+	for _, part := range parts {
+		result = append(result, strings.Replace(part, "_", "-", -1))
```

Comment:
Maybe also factor `strings.Replace(part, "_", "-", -1)` out to a function.

Reply:
Done, same for the other `string.Replace` above.